arch-design-questions.md

```diff
@@ -0,0 +1,93 @@
+# Client design
+
+## Principles
+
+- There are two top level clients - `ServiceBusReceiverClient` and `ServiceBusSenderClient`.
+- Clients are [single-use](#single-use-clients) - they are created with a receive mode, session ID (if applicable), queue/topic-subscription
+  baked in and cannot be changed.
+- The user can choose between two different methods of message delivery - push (via registered message handlers)
+  or push (via an iterator/reactor pattern or .receiveBatch()-style)
+- Sessions have two different methods for interacting with them:
+  1. [single-use](#single-use-clients)
+  2. [session manager](#sessions-design)
+- Different message types will be used for sending messages vs receiving messages. This is less confusing for
+  the user since we can remove fields that are non-sensical for sending (for instance, `locktoken`).
+
+# Single use clients
+
+Clients are single-use - you cannot change their receive mode or targeted queue/topic/subscription after
+they are created.
+
+This also means that calling .receive() (or registering multiple handlers) isn't a useful activity - the user
+would only be competing against _themselves_ since calling receive() multiple times would just result
```

Comment:
I'm not sure I'd say this as strongly; perhaps "calling receive concurrently isn't a useful activity"; since in python at least when we do "for message in client.receive()" we can break out, then call receive again.

Reply:
@KieranBrantnerMagee  - if they break out does that implicitly close the previous link? Or do you just close it if they call receive() again?

Reply:
Not sure that `concurrently` here fills in the mental picture for me.   It makes me think "oh, there's a race condition somewhere" rather than "don't do this multiple times for the same client"

Reply:
This actually is a race condition, although it's subtle.

The problem is if the user creates multiple receive links to the same entity and doesn't realize it. For instance, let's assume user does this:

const iter = client.getIterator()
iter.gottaIterate();
"abandon the iterator (but it's not actually closed)"

// ie, new receiver link
const newIter = client.getIterator();
newIter.gottaIterate()   // this is now "competing" with the other link which isn't even being monitored.

With eventhubs (minus checkpointing) this would have just resulted in multl-reading. With ServiceBus, thanks to it's server-side tracking you can just end up losing messages (receiveAndDelete mode) or possibly just arbitrarily delaying (peekLock w/o settlement) (assuming the previous receiver keeps trying to receive some messages based on whatever credits it threw on the line).

So when I say "not useful" I mean - you either compete with yourself (same TCP link, guessing it's more or less the same perf if you just dispatched from a single iterator loop) _or_ you accidentally create a race-condition and start obliterating messages (receiveAndDelete) or hiding them temporarily (peekLock) without realizing it.

Reply:
Could you lose messages, though?   If one link has a lock on a message, that would inherently prevent the second link of taking ownership, no?   I would expect the client to gracefully handle that case or actively prevent it by ensuring that there is only one link open to a given entity.

Reply:
You could lose messages if using prefetch and not actually inspecting the messages but this is independent of whether you have multiple iterators, I believe.

Reply:
Can you elaborate?   I'm not sure that I'm understanding the scenario and have the feeling that I'm overlooking something.

With the service managing locks and visibility, I can see the race condition where two links may receive duplicate messages due to prefetch, but once you begin to operate on them, the service should guard against settling a message that has already been settled.  Likewise with taking a lock on a message already locked.   I'm having difficulty coming up with a scenario where the consumers of both links would fail to have at least one be given the opportunity to process a message.

Reply:
If you are using prefetch, ReceiveAndDelete mode, and not actually receiving the messages, you would lose the messages that are prefetched.

Reply:
I was just mentioning this as the one way I know of where you can lose messages. It is totally independent of the multiple receive links scenario. As far as I know, there isn't an issue inherent to multiple receive links that would cause messages to be lost.

Reply:
GitHub decided that I'm not me and couldn't comment... so my follow-up was actually sitting unposted.   I think you just answered it.  :smile:

    Apologies; I'm not trying to be difficult but still don't think that I'm following your line of thought.   When you say "Not actually receiving the messages..."    
    Are you thinking of "these are on the prefetch queue but nobody has grabbed them to process" or something else?

Reply:
Correct, if you specify a prefetch amount greater than zero in your options bag, but you don't actually attempt to receive them, and you are in Receive And Delete mode, once your connection is closed, the messages will just be gone.

Reply:
While that scares me and is definitely worthy of at least a big bold call out in documentation,  that doesn't appear to be limited to multiple receiver connections to the same entity.  Unless I'm being dense, that's a not-awesome byproduct of prefetch.

Reply:
Absolutely, which I was attempting to get at in my previous comment. This is not related to multiple receive links. Sorry for any confusion!

Reply:
Ha, just read all of this! Never have so few sentences generated so much commentary!

I'll remove it since it's not adding much value anyways, but the discussion was interesting. :)